### PR TITLE
Gutenberg: Add a HMR & React Fast Refresh development mode.

### PR DESCRIPTION
**Pull request by epiqueras**
## Description

Closes #23013.

### Background

Webpack Hot Module Replacement (HMR) is fundamentally at odds with the way we do script overriding in Gutenberg.

After lots of experimentation and failed attempts, I found that the best way to support it is to provide an alternative development mode on a per-page/app basis. For example, the site editor is its own page/app, and this PR enables an HMR mode for its development.

It should be trivial to reuse the same configurations for the post, widgets, and navigation editors.

I also included the new [React Fast Refresh](https://reactnative.dev/docs/fast-refresh) experience. The developer experience with this is fantastic. Even error resilience and recovery are flawless, both for runtime and syntax errors. It's nice to be able to iterate on a screen without resetting state.

### Implementation

This PR introduces a new Webpack config for running the new dev server as an extension of the one provided by `@wordpress/scripts`.

Pages that wish to support an HMR mode, need to create a `hot.js` entry point like this PR does for `edit-site`. This entry point should initialize the page and load its root stylesheet.

On the server, we use a new config global, `GUTENBERG_HMR`, to determine whether to load the new script from the dev server or not. We manipulate this config variable effortlessly in the scripts introduced by this PR by using `wp-env`. It will enable it before running the new development mode, and disable it before running the old one. This makes it easy to switch between the two, but it does assume you are using `wp-env` to manage your environment.

## How to test this?

- Set up your environment using [`wp-env`](https://www.npmjs.com/package/@wordpress/env) if you haven't yet.
- Run `npm run dev:edit-site`.
- Go to the site editor and see how it updates in response to your code changes without reloading.

## Screenshots

gif

## Checklist:

- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: webpack.config.dev.js [hunk 1]
```diff
@@ -0,0 +1,67 @@
+/**
```
Comment (youknowriad) [inline webpack.config.dev.js:1]: Is there any way we can bundle this in wp-scripts and offer support for third-party plugin devs too?

**Comment by ItsJonQ** [comment]
@epiqueras Absolutely amazing!! I tested this locally, and it worked as expected!

* Used `wp-env` globally to `wp-env start`
* Ran `npm run dev:edit-site`
* Visit https://localhost:8888
* Enabled Full site editing in Gutenberg experiments
* Go to Site Editor
* Change files
* See the change ✨

File: webpack.config.dev.js (hunk 1 above)
Comment (epiqueras) [reply webpack.config.dev.js:1]: Definitely, I just wanted to stabilize it first.

Should I bundle it in in this PR or wait for more people who develop Gutenberg to use it and report potential issues?

Comment (youknowriad) [reply webpack.config.dev.js:1]: Ok, sounds like a reasonable plan to me.  

cc @gziolo for awareness too.

Comment (gziolo) [reply webpack.config.dev.js:1]: Nice one. It’s super exciting to see that it can be integrated into Gutenberg 👍

**Comment by chrisvanpatten** [comment]
I'm super excited about this, but I couldn't get it to work; I was getting "connection refused" errors on localhost:8081 so the editor couldn't load the file. I couldn't find any evidence that the port had already been allocated to another process either.

There were some warnings related to exports from Reakit not being found, but I'm not sure if those would be preventing the server from starting up or if it could be something else.

**Comment by epiqueras** [comment]
That's probably why. Delete node modules and reinstall.

**Comment by chrisvanpatten** [comment]
```
That's probably why. Delete node modules and reinstall.
```

I've done that three times now, once via `npm ci` and twice by deleting the folder. Also did `npm cache clear --force` and still nothing.

Here's the full set of errors…

```
WARNING in ./packages/components/build-module/alignment-matrix-control/index.js 82:23-32
"export 'Composite' was not found in 'reakit'
 @ ./packages/components/build-module/index.js
 @ ./packages/edit-site/src/plugins/index.js
 @ ./packages/edit-site/src/index.js
 @ ./packages/edit-site/hot.js
 @ multi ./node_modules/@pmmmwh/react-refresh-webpack-plugin/src/runtime/ReactRefreshEntry.js ./node_modules/@pmmmwh/react-refresh-webpack-plugin/src/runtime/ErrorOverlayEntry.js?sockPort=8081 ./packages/edit-site/hot.js

WARNING in ./packages/components/build-module/alignment-matrix-control/index.js 89:25-39
"export 'CompositeGroup' was not found in 'reakit'
 @ ./packages/components/build-module/index.js
 @ ./packages/edit-site/src/plugins/index.js
 @ ./packages/edit-site/src/index.js
 @ ./packages/edit-site/hot.js
 @ multi ./node_modules/@pmmmwh/react-refresh-webpack-plugin/src/runtime/ReactRefreshEntry.js ./node_modules/@pmmmwh/react-refresh-webpack-plugin/src/runtime/ErrorOverlayEntry.js?sockPort=8081 ./packages/edit-site/hot.js

WARNING in ./packages/components/build-module/alignment-matrix-control/cell.js 30:19-32
"export 'CompositeItem' was not found in 'reakit/Composite'
 @ ./packages/components/build-module/alignment-matrix-control/index.js
 @ ./packages/components/build-module/index.js
 @ ./packages/edit-site/src/plugins/index.js
 @ ./packages/edit-site/src/index.js
 @ ./packages/edit-site/hot.js
 @ multi ./node_modules/@pmmmwh/react-refresh-webpack-plugin/src/runtime/ReactRefreshEntry.js ./node_modules/@pmmmwh/react-refresh-webpack-plugin/src/runtime/ErrorOverlayEntry.js?sockPort=8081 ./packages/edit-site/hot.js

WARNING in ./packages/components/build-module/alignment-matrix-control/index.js 66:18-35
"export 'useCompositeState' was not found in 'reakit'
 @ ./packages/components/build-module/index.js
 @ ./packages/edit-site/src/plugins/index.js
 @ ./packages/edit-site/src/index.js
 @ ./packages/edit-site/hot.js
 @ multi ./node_modules/@pmmmwh/react-refresh-webpack-plugin/src/runtime/ReactRefreshEntry.js ./node_modules/@pmmmwh/react-refresh-webpack-plugin/src/runtime/ErrorOverlayEntry.js?sockPort=8081 ./packages/edit-site/hot.js

WARNING in ./packages/components/build-module/alignment-matrix-control/index.js 110:29-46
"export 'useCompositeState' was not found in 'reakit'
 @ ./packages/components/build-module/index.js
 @ ./packages/edit-site/src/plugins/index.js
 @ ./packages/edit-site/src/index.js
 @ ./packages/edit-site/hot.js
 @ multi ./node_modules/@pmmmwh/react-refresh-webpack-plugin/src/runtime/ReactRefreshEntry.js ./node_modules/@pmmmwh/react-refresh-webpack-plugin/src/runtime/ErrorOverlayEntry.js?sockPort=8081 ./packages/edit-site/hot.js
ℹ ｢wdm｣: Compiled with warnings.
```

**Comment by david-szabo97** [comment]
I was looking for HMR in Gutenberg but sadly found nothing. This would make the development a lot better!

* I had the same problem as Chris, but running `npm run build:packages` before `npm run dev:edit-site` did the trick.
* I also noticed that `npm run dev:edit-site` only rebuilds when something in `/packages/edit-site` changed. Can we get this working with all the modules? For example, if I make a change in `packages/block-editor/src/components/inserter/menu.js` then it should trigger the build. A quick hack that worked for me is running `npm run dev` and `npm run dev:edit-site` concurrently. First wait for `npm run dev` to finish then start the `npm run dev:edit-site` in another instance.

**Comment by noahtallen** [comment]
Superseded by #26922

**Pull request closed without merging**